druid-shell/src/platform/x11/application.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! X11 implementation of features at the application scope.
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use super::clipboard::Clipboard;
+use super::window::XWindow;
+use crate::application::AppHandler;
+use crate::kurbo::Point;
+use crate::{KeyCode, KeyModifiers, MouseButton, MouseEvent};
+
+struct XcbConnection {
+    connection: Arc<xcb::Connection>,
+    screen_num: i32,
+}
+
+lazy_static! {
+    static ref XCB_CONNECTION: XcbConnection = XcbConnection::new();
+}
+
+thread_local! {
+    static WINDOW_MAP: RefCell<HashMap<u32, XWindow>> = RefCell::new(HashMap::new());
+}
+
+pub struct Application;
+
+impl Application {
+    pub fn new(_handler: Option<Box<dyn AppHandler>>) -> Application {
+        Application
+    }
+
+    pub(crate) fn add_xwindow(id: u32, xwindow: XWindow) {
+        WINDOW_MAP.with(|map| map.borrow_mut().insert(id, xwindow));
+    }
+
+    pub(crate) fn get_connection() -> Arc<xcb::Connection> {
+        XCB_CONNECTION.connection_cloned()
+    }
+
+    pub(crate) fn get_screen_num() -> i32 {
+        XCB_CONNECTION.screen_num()
+    }
+
+    pub fn run(&mut self) {
+        let conn = XCB_CONNECTION.connection_cloned();
+        loop {
+            if let Some(ev) = conn.wait_for_event() {
+                let ev_type = ev.response_type() & !0x80;
+                //NOTE: I don't think we should be doing this here, but I'm trying to keep
+                //the code mostly unchanged. My personal feeling is that the best approach
+                //is to dispatch events to the window as early as possible, that is to say
+                //I would send the *raw* events to the window and then let the window figure
+                //out how to handle them. - @cmyr
+                match ev_type {
```

Comment:
so to expand on this comment: I think we should make this runloop as minimal as possible, and basically dispatch raw events to our internal window object with as little handling as possible, and do all this logic there.

In particular I feel like this will make it easier to eventually support the VST case; the window object can basically stand alone.

Reply:
Well -- to get which window we need to send the event to, we actually need to parse the event within this loop and fetch the window ID out of it (`expose.window()`, `key_press.event()`, `button_press.event()` etc)

We *could* send the raw event and let the window parse the event again, though, in the spirit of handling everything in the window -- it's probably not too much overhead to parse twice.

Reply:
Hmm, okay. Another option I might consider is to have our own `XcbEvent` enum; we would parse these raw events into that type here, extract the window id, and then forward the event to the window?

I'm just spitballing, really, I don't have super-strong opinions.

Reply:
I'm thinking it's very possible that I'll have to work in this area when I'm exploring VST compatibility, so I'm leaning towards re-thinking this then? Agree that it's not ideal, though.